File: 290. Word Pattern/main.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

using std::getline;
using std::istringstream;
using std::string;
using std::stringstream;
using std::unordered_map;
using std::vector;

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        istringstream iss(s);
        string word;
        char deimiter = ' ';
        while (getline(iss, word, deimiter)) {
            if (!word.empty()) {
                words.push_back(word);
            }
        }

        size_t patternLen = pattern.size();
        size_t wordsLen = words.size();
        if (patternLen != wordsLen) {
            return false;
        }

        unordered_map<char, string> directMapping;
        unordered_map<string, char> reverseMapping;

        for (size_t i = 0; i < patternLen; i++) {
            if ((directMapping.find(pattern[i]) != directMapping.end() && directMapping[pattern[i]] != words[i]) ||
                (reverseMapping.find(words[i]) != reverseMapping.end() && reverseMapping[words[i]] != pattern[i])) {
                return false;
            }

            directMapping[pattern[i]] = words[i];
            reverseMapping[words[i]] = pattern[i];
        }

        return true;
    }
};

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        istringstream iss(s);
        string word;

        while (iss >> word) {
            words.push_back(word);
        }

        size_t patternLen = pattern.size();
        size_t wordsLen = words.size();
        if (patternLen != wordsLen) {
            return false;
        }

        unordered_map<char, string> directMapping;
        unordered_map<string, char> reverseMapping;

        for (size_t i = 0; i < patternLen; i++) {
            if ((directMapping.find(pattern[i]) != directMapping.end() && directMapping[pattern[i]] != words[i]) ||
                (reverseMapping.find(words[i]) != reverseMapping.end() && reverseMapping[words[i]] != pattern[i])) {
                return false;
            }

            directMapping[pattern[i]] = words[i];
            reverseMapping[words[i]] = pattern[i];
        }

        return true;
    }
};

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        stringstream ss(s);
        string word;

        while (ss >> word) {
            words.push_back(word);
        }

        size_t patternLen = pattern.size();
        size_t wordsLen = words.size();
        if (patternLen != wordsLen) {
            return false;
        }

        unordered_map<char, string> directMapping;
        unordered_map<string, char> reverseMapping;

        for (size_t i = 0; i < patternLen; i++) {
            if ((directMapping.find(pattern[i]) != directMapping.end() && directMapping[pattern[i]] != words[i]) ||
                (reverseMapping.find(words[i]) != reverseMapping.end() && reverseMapping[words[i]] != pattern[i])) {
                return false;
            }

            directMapping[pattern[i]] = words[i];
            reverseMapping[words[i]] = pattern[i];
        }

        return true;
    }
};

class Solution {
    static constexpr size_t ARRAY_SIZE = 26;

public:
    bool wordPattern(string pattern, string s) {
        size_t patternLen = pattern.size();
        vector<string> hashmap(ARRAY_SIZE, "");
        stringstream ss(s);
        string word;
        size_t idx = 0;

        while (ss >> word) {
            if (idx >= patternLen) {
                return false;
            }

            size_t index = pattern[idx] - 'a';
            if (hashmap[index].empty()) {
                for (size_t i = 0; i < ARRAY_SIZE; i++) {
                    if (!hashmap[i].empty() && hashmap[i] == word) {
                        return false;
                    }
                }
                hashmap[index] = word;
            } else {
                if (hashmap[index] != word) {
                    return false;
                }
            }

            idx++;
        }

        return idx == patternLen;
    }
};

File: 290. Word Pattern/main.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")
        patternLen = len(pattern)
        wordsLen = len(words)

        if patternLen != wordsLen:
            return False

        directMapping = dict()
        reverseMapping = dict()

        for i in range(patternLen):
            if (
                pattern[i] in directMapping and directMapping[pattern[i]] != words[i]
            ) or (
                words[i] in reverseMapping and reverseMapping[words[i]] != pattern[i]
            ):
                return False

            directMapping[pattern[i]] = words[i]
            reverseMapping[words[i]] = pattern[i]

        return True


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")
        patternLen = len(pattern)
        wordsLen = len(words)

        if patternLen != wordsLen:
            return False

        HASHMAP_SIZE = 26
        hashmap = [""] * HASHMAP_SIZE

        for i in range(patternLen):
            index = ord(pattern[i]) - ord("a")
            if hashmap[index]:
                if hashmap[index] != words[i]:
                    return False
            else:
                for j in range(HASHMAP_SIZE):
                    if hashmap[j] == words[i]:
                        return False
                hashmap[index] = words[i]

        return True


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")
        patternLen = len(pattern)
        wordsLen = len(words)

        if patternLen != wordsLen:
            return False

        HASHMAP_SIZE = 26
        hashmap = [""] * HASHMAP_SIZE

        for i in range(patternLen):
            index = ord(pattern[i]) - ord("a")
            if hashmap[index]:
                if hashmap[index] != words[i]:
                    return False
            else:
                if words[i] in hashmap:
                    return False
                hashmap[index] = words[i]

        return True